src/garage/envs/garage_env.py

```diff
@@ -162,7 +162,8 @@ def step(self, action):
         # 'GarageEnv.TimeLimitTerminated'
         if 'TimeLimit.truncated' in info:
             info['GarageEnv.TimeLimitTerminated'] = done  # done = True always
-            done = not info['TimeLimit.truncated']
+            if info['TimeLimit.truncated']:
```

Comment:
i don't understand why this change is needed.

if `info['TimeLimit.truncated'] = False` and `done = True` then the environment terminated before the horizon limit and we should pass along `done = True`, i.e. `done = done and not info['TimeLimit.truncated']`

i don't think this configuration can actually appear in the current implementation of gym (which only ever populates `info['TimeLimit.truncated'] = True` at the horizon limit and omits it otherwise, but it is logically valid.

Reply:
> If you are going to fix a ton of issues in one PR, please document all of them extensively in the commit message. Right now, the PR title is misleading.

Sorry I'll take care of that.

Reply:
@ryanjulian afaict this isn't the case.

The gym time limit truncated wrapper will only ever populate a info['TimeLimit.truncated'] = False iff the environment has been stepped past its max episode length AND done = True.

Another way of phrasing this is that timelimit truncated envs only ever return info['TimeLimit.truncated'] = True when the env has been stepped past its max episode length AND done = False.

See this line inside the timelimit gym wrapper for where this logic is implemented.
https://github.com/openai/gym/blob/abb815c871447a14532558772739ccb88deee109/gym/wrappers/time_limit.py#L19

Reply:
simply unwinding the logic from the code you linked, recovering the original value of `done`, looks like this:

```python
if 'TimeLimit.truncated' in info:
    info['GarageEnv.TimeLimitTerminated'] = True
    done = not info['TimeLimit.truncated']
else:
    info['GarageEnv.TimeLimitTerminated'] = False
    info['TimeLimited.truncated'] = False
```

Why would you change it to be unconditional of the value of `info['TimeLimit.truncated']`? Nothing in the code you linked implies that `done` MUST be `False` when `self._elapsed_steps >= self._max_episode_steps`

Reply:
@ryanjulian well a problem comes if we wrap an environment this way: `GarageEnv(MultiEnvwrapper([GarageEnv(env)]))`, which is what happens in all of our MT examples.

Assuming an environment hasn't hit a timelimit here's what ends up happening is the following:

The environment gets stepped and inside the `MultiEnvWrapper` the first `GarageEnv` wrapper adds the info entry `info['TimeLimit.truncated'] = False`. Then the outer `GarageEnv` wrapper evaluates the info that is returned by the inner `GarageEnv` wrapper, which has info with the entry `info['TimeLimit.truncated'] = False`. The outer `GarageEnv` Wrapper now enters the if statement that you linked above:
```python
if 'TimeLimit.truncated' in info:
    info['GarageEnv.TimeLimitTerminated'] = True
    done = not info['TimeLimit.truncated']
```
it now sets `info['GarageEnv.TimeLimitTerminated'] = True` and `done = not False` which is the incorrect behavior. `Done` should only be modified when `info['TimeLimit.truncated'] = True` never when it is `False`.

That is the logic behind why I made this change.

Reply:
isn't the bug then that it was wrapped twice with `GarageEnv`? what should make me believe this won't break the non-pathological case where there's only one `GarageEnv`?

Reply:
why not just detect double-wrapping and then vomit an error?

Reply:
> isn't the bug then that it was wrapped twice with `GarageEnv`?

I thought this was a valid use case, which is why I wrote `MultiEnvWrapper` the way that I did with nested `GarageEnvs`

> what should make me believe this won't break the non-pathological case where there's only one `GarageEnv`

There are 3 input cases here afaict if we assume that we are handling only the output of the `TimeLimitWrapper`:

1) `'TimeLimit.truncated' = True and done = True`
	
	we enter the proposed if statement `if info['TimeLimit.truncated']:` and set `done = False`
2) `'TimeLimit.truncated' = False and done = True`

	we don't enter the newly proposed if statement, and done is not modified.
3) `'TimeLimit.truncated'` isn't in `env_infos`.
	we don't enter the newly proposed if statement, and done is not modified.

Reply:
If it's a valid use case, it's extremely strange and surprising to users -- why would I wrap the same thing twice? Is wrapping twice strictly necessary in the first place?

In any case, it's apparently *not* valid to wrap `GarageEnv` twice, because apparently `GarageEnv` isn't idempotent. So either you need to make `GarageEnv` idempotent, or you need to detect double-wrapping and throw an error.

Here's the truth table as I read it:
| `'TimeLimited.truncated' in info` | `info['TimeLimit.truncated']` | done  | done = |
|---|---|---|---|
| `False` | N/A | `False` | `False` |
| `False` | N/A | `True` | `True` |
| `True` | `False` | `False` | `True` |
| `True` | `False` | `True` | `True` |
| `True` | `True` | `False` | `False` |
| `True` | `True` | `True` | `False` |

As long as your backport implements this truth table for arbitrary nestings of `GarageEnv` (or only one wrapping of `GarageEnv`, and throws an error if you do it twice), I am fine with it.

Reply:
If it's a valid use case, it's really strange. What about the definition of `GarageEnv` makes it seem like it would be intended to be used twice? That would be really surprising to any user.

In any case, it's apparently *not* a valid use case, because `GarageEnv` is apparently not idempotent. So you *either* need to make `GarageEnv` idempotent, *or* you can leave it non-idempotent, but detect double-wrapping and throw an error.

Here's the truth table as I see it:
| `'TimeLimit.truncated' in info` | `info['TimeLimit.truncated']` | `done` | `done = ` |
|---|---|---|---|
| `False` | N/A | `False` | `False` |
| `False` | N/A | `True` | `True` |
| `True` | `False` | `False` | `True` |
| `True` | `False` | `True` | `True` |
| `True` | `True` | `False` | `False` |
| `True` | `True` | `True` | `False` |

As long as your fix implements this truth table for arbitrary nestings of GarageEnv (or only one wrapping, but throws errors on double-wrapping), I'm happy.

Please prove it with a test, since obviously this is getting pretty subtle.

Reply:
gotcha, yes it implements this change. Let me go ahead and implement a test that checks this on master and on this release branch.